File: generador.py
import sys

class Automata:
    def __init__(self, estados, alfabeto, transiciones, estado_inicial, estados_finales):
        self.estados = estados
        self.alfabeto = alfabeto
        
        # Validar origen de transiciones
        if not all(origen in estados for (origen, letra) in transiciones.keys()):
            print("Las transiciones deben estar con estados válidos")
            sys.exit()

        self.transiciones = transiciones

        if estado_inicial not in estados:
            print("El estado inicial no está en el conjunto")
            sys.exit()

        self.estado_inicial = estado_inicial

        if not all(estado in estados for estado in estados_finales):
            print("Los estados finales no están en el conjunto")
            sys.exit()

        self.estados_finales = estados_finales
        

def main():
    if len(sys.argv) != 2:
        print("Agregue un archivo")
        return

    caracteres = str.maketrans("", "", "{}\n()")
    nombre_archivo = sys.argv[1]

    try:
        with open(nombre_archivo, "r") as f:
            lineas = f.readlines() 
       
            if len(lineas) != 5:
                print("El archivo debe tener 5 líneas")
                return

            estados = lineas[0].translate(caracteres).split(",")
            alfabeto = lineas[1].translate(caracteres).split(",")
            estados = [e.strip() for e in lineas[0].translate(caracteres).split(",")]
            alfabeto = [a.strip() for a in lineas[1].translate(caracteres).split(",")]

            transiciones_raw = lineas[2].translate(caracteres).split(",")
            transiciones = {}
            print(transiciones_raw)
            for t in transiciones_raw:
                if not t.strip():
                    continue

                izquierda, destino = t.split("=", 1)

                izquierda = izquierda.split(":")

                transiciones[(izquierda[0], izquierda[1])] = destino

            estado_inicial = lineas[3].translate(caracteres).strip()
            estados_finales = lineas[4].translate(caracteres).split(",")
            automata = Automata(estados, alfabeto, transiciones, estado_inicial, estados_finales)
            print("Automata creado correctamente")
            print("Estados:", automata.estados)
            print("Transiciones:", automata.transiciones)

    except FileNotFoundError:
        print("Archivo no encontrado.")
    except Exception as e:
        print(f"Ocurrio un error: {e}")


if __name__ == "__main__":
    main()

File: generador.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX_STATES 50
#define MAX_ALPHABET 50
#define MAX_TRANSITIONS 250
#define MAX_LINE_LENGTH 500
#define MAX_STRING_LENGTH 250

// Estructura para representar una transicion
typedef struct {
    char estado_actual[MAX_STATES];
    char simbolo;
    char estado_siguiente[MAX_STATES];
} Transicion;

typedef struct {
    char estados[MAX_STATES][MAX_STATES];
    int num_estados;
    char alfabeto[MAX_ALPHABET];
    int num_simbolos;
    Transicion transiciones[MAX_TRANSITIONS];
    int num_transiciones;
    char estado_inicial[MAX_STATES];
    char estados_finales[MAX_STATES][MAX_STATES];
    int num_estados_finales;
} Automata;

// funciones
void leer_automata(const char* archivo_automata, Automata* af);
void procesar_cadenas(const char* archivo_cadenas, Automata* af);
int aceptar_cadena(const char* cadena, Automata* af);

// Función principal
int main(int argc, char* argv[]) {
    if (argc != 3) {
        printf("Uso: %s <automata.txt> <cadenas.txt>\n", argv[0]);
        return 1;
    }

    Automata af;
    leer_automata(argv[1], &af);
    procesar_cadenas(argv[2], &af);

    return 0;
}


void leer_estados(FILE* file, Automata* af) {
    char line[MAX_LINE_LENGTH];
    fgets(line, sizeof(line), file);
    line[strcspn(line, "\n")] = 0; // Eliminar el salto de línea

    char* token = strtok(line, ","); // Elimina las comas con la funcion strtok
    af->num_estados = 0;
    while (token != NULL) {
        strcpy(af->estados[af->num_estados++], token);
        token = strtok(NULL, ",");
    }
}

void leer_alfabeto(FILE* file, Automata* af) {
    char line[MAX_LINE_LENGTH];
    fgets(line, sizeof(line), file);
    line[strcspn(line, "\n")] = 0;

    char* token = strtok(line, ",");
    af->num_simbolos = 0;
    while (token != NULL) {
        af->alfabeto[af->num_simbolos++] = token[0];
        token = strtok(NULL, ",");
    }
}

// Lee la línea de transiciones
void leer_transiciones(FILE* file, Automata* af) {
    char line[MAX_LINE_LENGTH];
    fgets(line, sizeof(line), file);
    line[strcspn(line, "\n")] = 0;

    char* token = strtok(line, ",");
    af->num_transiciones = 0;
    while (token != NULL) {
        // Formato: (q0:a)=q1
        char from[MAX_STATES], to[MAX_STATES];
        char symbol;
        sscanf(token, "(%[^:]:%c)=%s", from, &symbol, to);

        strcpy(af->transiciones[af->num_transiciones].estado_actual, from);
        af->transiciones[af->num_transiciones].simbolo = symbol;
        strcpy(af->transiciones[af->num_transiciones].estado_siguiente, to);
        af->num_transiciones++;

        token = strtok(NULL, ",");
    }
}

// Lee la línea del estado inicial
void leer_estado_inicial(FILE* file, Automata* af) {
    char line[MAX_LINE_LENGTH];
    fgets(line, sizeof(line), file);
    line[strcspn(line, "\n")] = 0;
    strcpy(af->estado_inicial, line);
}

// Lee la línea de estados finales
void leer_estados_finales(FILE* file, Automata* af) {
    char line[MAX_LINE_LENGTH];
    fgets(line, sizeof(line), file);
    line[strcspn(line, "\n")] = 0;

    char* token = strtok(line, ",");
    af->num_estados_finales = 0;
    while (token != NULL) {
        strcpy(af->estados_finales[af->num_estados_finales++], token);
        token = strtok(NULL, ",");
    }
}


void leer_automata(const char* archivo_automata, Automata* af) {
    FILE* file = fopen(archivo_automata, "r");
    if (file == NULL) {
        perror("Error al abrir el archivo del automata");
        exit(1);
    }

    leer_estados(file, af);
    leer_alfabeto(file, af);
    leer_transiciones(file, af);
    leer_estado_inicial(file, af);
    leer_estados_finales(file, af);

    fclose(file);
}

// Procesa el archivo de cadenas
void procesar_cadenas(const char* archivo_cadenas, Automata* af) {
    FILE* file = fopen(archivo_cadenas, "r");
    if (file == NULL) {
        perror("Error al abrir el archivo de cadenas");
        exit(1);
    }

    char cadena[MAX_STRING_LENGTH];
    while (fgets(cadena, sizeof(cadena), file) != NULL) {
        cadena[strcspn(cadena, "\n")] = 0; // Eliminar el salto de línea
        if (aceptar_cadena(cadena, af)) {
            printf("La cadena '%s' es ACEPTADA.\n", cadena);
        } else {
            printf("La cadena '%s' es RECHAZADA.\n", cadena);
        }
    }

    fclose(file);
}

int aceptar_cadena(const char* cadena, Automata* af) {
    char estado_actual[MAX_STATES];
    strcpy(estado_actual, af->estado_inicial);

    for (int i = 0; i < strlen(cadena); i++) {
        char simbolo_actual = cadena[i];
        int transicion_encontrada = 0;
        char siguiente_estado[MAX_STATES] = "";

        for (int j = 0; j < af->num_transiciones; j++) {
            if (strcmp(af->transiciones[j].estado_actual, estado_actual) == 0 &&
                af->transiciones[j].simbolo == simbolo_actual) {
                strcpy(siguiente_estado, af->transiciones[j].estado_siguiente);
                transicion_encontrada = 1;
                break;
            }
        }

        if (!transicion_encontrada) {
            return 0;
        }
        strcpy(estado_actual, siguiente_estado);
    }

    // Verificar si el estado final es un estado de aceptación
    for (int i = 0; i < af->num_estados_finales; i++) {
        if (strcmp(estado_actual, af->estados_finales[i]) == 0) {
            return 1; // El estado actual es final, la cadena es aceptada
        }
    }
    return 0; // El estado final no es de aceptación, la cadena es rechazada
}